src/compiler/extensions/buildInstructions.js

```diff
@@ -5,60 +5,80 @@ import Instruction from '../utils/Instruction';
 
 let generateWalker = visitor.build({
   TORNADO_PARTIAL: {
-    enter(item, ctx) {
-      ctx.pushInstruction(new Instruction('insert', {key: item.node[1].key, item, ctx}));
+    enter(item, instructions, state) {
+      instructions.push(new Instruction('insert_TORNADO_PARTIAL', {key: item.node[1].key}, state));
     }
   },
   TORNADO_BODY: {
-    enter(item, ctx) {
-      ctx.pushInstruction(new Instruction('open', {key: item.node[1].key, item, ctx}));
+    enter(item, instructions, state) {
+      let params;
+      instructions.push(new Instruction('open_TORNADO_BODY', {key: item.node[1].key, type: item.node[1].type, name: item.node[1].name}, state));
+      params = item.node[1].params;
+      if (params && params.length) {
+        params.forEach(function(p) {
+          // we currently only support a string or number or a reference
+          // TODO: move this logic into the visitor
+          let val = p[1].val,
+              key = p[1].key;
+          instructions.push(new Instruction('open_TORNADO_PARAM', {key: key}, state));
+          if (typeof val === 'string') {
+            instructions.push(new Instruction('insert_PLAIN_TEXT', {content: val, type: 'plaintext'}, state));
+          } else if (typeof val === 'number') {
+            instructions.push(new Instruction('insert_PLAIN_TEXT', {content: val, type: 'plaintext'}, state));
+          } else {
+            instructions.push(new Instruction('insert_TORNADO_REFERENCE', {key: val[1].key}, state));
```

Comment:
Maybe we should call `this.TORNADO_REFERENCE.enter` here, so any changes to how `TORNADO_REFERENCE` instructions are created will be picked up here as well.